include/deal.II/lac/trilinos_tpetra_precondition.h

```diff
@@ -0,0 +1,1311 @@
+
+      /**
+       * Return the partitioning of the range space of this matrix, i.e., the
+       * partitioning of the vectors that are result from matrix-vector
+       * products.
+       *
+       * @return IndexSet of the range of this preconditioner
+       */
+      IndexSet
+      locally_owned_range_indices() const;
+
+      /** @} */
+
+      /**
+       * @addtogroup Exceptions
+       */
+      /** @{ */
+      /**
+       * The maps of the underlying matrix and one of the vectors
+       * do not match.
+       */
+      DeclException1(
+        ExcNonMatchingMaps,
+        std::string,
+        << "The sparse matrix the preconditioner is based on "
+        << "uses a map that is not compatible to the one in vector " << arg1
+        << ". Check preconditioner and matrix setup.");
+
+      /**
+       * The chosen preconditioner does not support a transposed apply.
+       */
+      DeclExceptionMsg(
+        ExcTransposeNotSupported,
+        "The chosen preconditioner does not support transposing the matrix.");
+      /** @} */
+
+    protected:
+      /**
+       * This is a pointer to the preconditioner object that is used when
+       * applying the preconditioner.
+       */
+      Teuchos::RCP<TpetraTypes::LinearOperator<Number, MemorySpace>>
+        preconditioner;
+
+      /**
+       * @brief The list of preconditioner parameters.
+       *
+       * This structure is Trilinos counterpart to the AdditionalData structures
+       * in deal.II. Therefore any initialize will at some point pass this to
+       * the preconditioner. Most derived classes will handle building this list
+       * based on an AdditionalData object that exposes and defaults the most
+       * sensible parameters. But some classes will also offer full
+       * customization for experienced Trilinos users.
+       *
+       */
+      Teuchos::ParameterList parameter_list;
+    };
+
+
+#  ifdef DEAL_II_TRILINOS_WITH_IFPACK2
+    /**
+     * @brief Wrapper class for the IdentitySolver preconditioner of Ifpack2.
+     *
+     * @ingroup TpetraWrappers
+     * @ingroup Preconditioners
+     *
+     * @tparam Number scalar type of the preconditioner
+     * @tparam MemorySpace
+     */
+    template <typename Number, typename MemorySpace = dealii::MemorySpace::Host>
+    class PreconditionIdentity : public PreconditionBase<Number, MemorySpace>
+    {
+    public:
+      /**
+       * @brief Construct Identity preconditioner.
+       *
+       */
+      PreconditionIdentity() = default;
+
+      /**
+       * @brief Initializes the preconditioner for the matrix <tt>A</tt>.
+       *
+       * Note that this only needs the matrix for information on the parallel
+       * distribution of the matrix and will return the original vector on
+       * <tt>vmult</tt> or <tt>Tvmult</tt>.
+       *
+       * @param A Matrix to base the preconditioner on.
+       */
+      void
+      initialize(const SparseMatrix<Number, MemorySpace> &A);
+    };
+
+    /**
```

Comment:
Here and elsewhere, could you please uniformly insert 3 empty lines between class definitions. We are not always strictly following this, but we recommend it at https://www.dealii.org/developer/doxygen/deal.II/CodingConventions.html (point 4) so it might be worth considering.

Reply:
Done!